components/places/src/bookmark_sync/tests/payload_evolution.rs

```diff
@@ -0,0 +1,318 @@
+            remote_bookmark(guid('b'), "Cats", vec![]),
+            remote_folder(
+                guid('c'),
+                "Animals",
+                vec!["A", "B"],
+                vec![("device_type", "phone"), ("color", "blue")],
+            ),
+            remote_separator(guid('d'), 2, vec![("device_type", "desktop")]),
+            // Create the unfiled bookmarks folder, or else dogear will try to delete the query
+            // rather than sync it (https://github.com/mozilla/application-services/pull/5438#discussion_r1143960257)
+            remote_item(
+                json!({
+                    "id": "unfiled",
+                    "title": "Unfiled Bookmarks",
+                    "type": "folder",
+                    "parentid": "places",
+                    "parentName": "places",
+                    "dateAdded": 0,
+                    "children": vec![guid('e')],
+                }),
+                vec![],
+            ),
+            // Pretend like we can update query items to test payload evolution, even though we
+            // don't actually support it in the API.
+            remote_query(
+                guid('e'),
+                "dog search",
+                "place:folder=123&excludeItems=1",
+                vec![("device_type", "desktop")],
+            ),
+        ],
+        local_updates: vec![
+            title_change(guid('a'), "Doggies"),
+            title_change(guid('b'), "Kitties"),
+            title_change(guid('c'), "Cute Animals"),
+            pos_change(guid('d'), 3),
+            title_change(guid('e'), "Doggy search"),
+        ],
+        incoming_remote_records: vec![],
+        outgoing_unknown_fields: vec![
+            (guid('a'), vec![("device_type", "phone")]),
+            (guid('b'), vec![]),
+            (guid('c'), vec![("device_type", "phone"), ("color", "blue")]),
+            (guid('d'), vec![("device_type", "desktop")]),
+            (guid('e'), vec![("device_type", "desktop")]),
+        ],
+    }
+    .run()
+}
+
+#[test]
+fn test_new_unknown_fields() {
+    // If we have new incoming remote records with new unknown fields, those should override the
+    // ones from the mirror table
+    RoundtripTest {
+        initial_remote_records: vec![
+            remote_bookmark(guid('a'), "Dogs", vec![("device_type", "phone")]),
+            remote_bookmark(guid('b'), "Cats", vec![("device_type", "desktop")]),
+        ],
+        local_updates: vec![
+            title_change(guid('a'), "Doggies"),
+            title_change(guid('b'), "Kitties"),
+        ],
+        incoming_remote_records: vec![remote_bookmark(
+            guid('a'),
+            "Dogs",
+            vec![("device_type", "mini-phone")],
+        )],
+        outgoing_unknown_fields: vec![
+            (guid('a'), vec![("device_type", "mini-phone")]),
+            (guid('b'), vec![("device_type", "desktop")]),
+        ],
+    }
+    .run()
+}
+
+struct RoundtripTest {
```

Comment:
❤️ table-driven tests!

Reply:
Yeah, they're pretty great.  I saw a few in the app-services code when I first came on and copied the style.  The more I write the more I like them.